controllers/awsmanagedcontrolplane_controller.go

```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		logger.Error(err, "Failed to get aws client session")
+		return ctrl.Result{}, microerror.Mask(err)
+	}
+
+	var iamService *iam.IAMService
+	{
+		c := iam.IAMServiceConfig{
+			AWSSession:       awsClientSession,
+			ClusterName:      clusterName,
+			MainRoleName:     *eksCluster.Spec.RoleName,
+			Log:              logger,
+			RoleType:         iam.IRSARole,
+			Region:           eksCluster.Spec.Region,
+			IAMClientFactory: r.IAMClientFactory,
+		}
+		iamService, err = iam.New(c)
+		if err != nil {
+			logger.Error(err, "Failed to generate IAM service")
+			return ctrl.Result{}, microerror.Mask(err)
+		}
+	}
+
+	if eksCluster.DeletionTimestamp != nil {
+		err = iamService.DeleteRolesForIRSA()
+		if err != nil {
+			return ctrl.Result{}, microerror.Mask(err)
+		}
+		// remove finalizer from AWSManagedControlPlane
+		{
+			if controllerutil.ContainsFinalizer(eksCluster, key.FinalizerName(iam.IRSARole)) {
+				patchHelper, err := patch.NewHelper(eksCluster, r.Client)
+				if err != nil {
+					return ctrl.Result{}, microerror.Mask(err)
+				}
+				controllerutil.RemoveFinalizer(eksCluster, key.FinalizerName(iam.IRSARole))
+				err = patchHelper.Patch(ctx, eksCluster)
+				if err != nil {
+					logger.Error(err, "failed to remove finalizer on AWSManagedControlPlane")
+					return ctrl.Result{}, microerror.Mask(err)
+				}
+				logger.Info("successfully removed finalizer from AWSManagedControlPlane", "finalizer_name", iam.IRSARole)
+			}
+		}
+	} else {
+		// add finalizer to AWSManagedControlPlane
+		if !controllerutil.ContainsFinalizer(eksCluster, key.FinalizerName(iam.IRSARole)) {
+			patchHelper, err := patch.NewHelper(eksCluster, r.Client)
+			if err != nil {
+				return ctrl.Result{}, microerror.Mask(err)
+			}
+			controllerutil.AddFinalizer(eksCluster, key.FinalizerName(iam.IRSARole))
+			err = patchHelper.Patch(ctx, eksCluster)
+			if err != nil {
+				logger.Error(err, "failed to add finalizer on AWSManagedControlPlane")
+				return ctrl.Result{}, microerror.Mask(err)
+			}
+			logger.Info("successfully added finalizer to AWSManagedControlPlane", "finalizer_name", iam.IRSARole)
+		}
+
+		accountID, err := getAWSAccountID(awsClusterRoleIdentity)
+		if err != nil {
+			logger.Error(err, "Could not get account ID")
+			return ctrl.Result{}, microerror.Mask(err)
+		}
+
+		eksOpenIdDomain, err := iamService.GetIRSAOpenIDForEKS(eksCluster.Name)
+		if err != nil {
+			logger.Error(err, "failed to fetch EKS OpenConnectID URL")
+			return ctrl.Result{}, microerror.Mask(err)
+		}
+
+		eksRoleARN, err := iamService.GetRoleARN(*eksCluster.Spec.RoleName)
+		if err != nil {
+			logger.Error(err, "failed to fetch EKS role name ARN")
+
+			return ctrl.Result{}, microerror.Mask(err)
+		}
+
+		iamService.SetPrincipalRoleARN(eksRoleARN)
+		err = iamService.ReconcileRolesForIRSA(accountID, eksOpenIdDomain)
+		if err != nil {
+			return ctrl.Result{}, microerror.Mask(err)
+		}
+	}
+
+	return ctrl.Result{
+		Requeue: true,
```

Comment:
This will always requeue, normally reconciling the CR again immediately. It sounds like this will hammer the api quiet a bit, can we requeue every X minutes instead?

Reply:
I was under impression that this will mean requeue in the default requeue time if that mean immediately than this is wrong for sure, i will add time